.github/workflows/publish.yml

```diff
@@ -1,63 +1,110 @@
 name: publish
 on:
   push:
-    branches:
+    branches: # For branches, better to list them explicitly than regexp include
       - master
-      - "[0-9].[0-9]+.x"
+      - 3.3.x
 jobs:
-  checks:
-    name: publication
+  # General job notes: we DON'T want to cancel any previous runs, especially in the case of a "back to snapshots" build right after a release push
+  prepare:
+    # Notes on prepare: this job has no access to secrets, only github token. As a result, all non-core actions are centralized here
+    # This includes the tagging and drafting of release notes. Still, when possible we favor plain run of gradle tasks
+    name: prepare
     runs-on: ubuntu-latest
+    outputs:
+      versionType: ${{ steps.version.outputs.versionType }}
     steps:
-    # we DON'T want to cancel previous runs, especially in the case of a "back to snapshots" build right after a release push
     - uses: actions/checkout@v2
-    - uses: christian-draeger/read-properties@1.0.1
-      name: read version
-      id: read-version
+    - name: setup java
+      uses: actions/setup-java@v1
       with:
-        path: './gradle.properties'
-        property: 'version'
-    - uses: frabert/replace-string-action@v1.2
-      id: split-qualifier
-      with:
-        string: ${{ steps.read-version.outputs.value }}
-        pattern: '(\d+\.\d+\.\d+)[.-]?([\w-]*)'
-        replace-with: 'VERSION$2'
+        java-version: 8
     - name: interpret version
-      id: version-type
-      run: |
-          echo "::set-output name=issnapshot::${{ steps.split-qualifier.outputs.replaced == 'VERSIONSNAPSHOT' || steps.split-qualifier.outputs.replaced == 'VERSIONBUILD-SNAPSHOT'  }}"
-          echo "::set-output name=isprerelease::${{ startsWith(steps.split-qualifier.outputs.replaced,'VERSIONRC') || startsWith(steps.split-qualifier.outputs.replaced,'VERSIONM')  }}"
-          echo "::set-output name=isrelease::${{ steps.split-qualifier.outputs.replaced == 'VERSIONRELEASE' || steps.split-qualifier.outputs.replaced == 'VERSION'  }}"
-          echo "::set-output name=isunknown::${{ startsWith(steps.split-qualifier.outputs.replaced,'VERSION') != true  }}"
-    - name: validate version interpretation
-      if: steps.version-type.outputs.isunknown == 'true'
+      id: version
+      #we only run the qualifyVersionGha task so that no other console printing can hijack this step's output
+      #output: versionType, fullVersion
+      run: ./gradlew qualifyVersionGha
+    - name: fail-if-bad
```

Comment:
idea: if you fail the execution of `qualifyVersionGha` task, this step won't be needed

Reply:
updated